File: python/day25.py
import unittest
import time


def solve_part_1(public_card: int, public_door: int) -> int:
    loop_size = 0
    value = 1
    while value != public_card:
        value *= 7
        value %= 20201227
        loop_size += 1

    return pow(public_door, loop_size, 20201227)


class Testing(unittest.TestCase):

    test_input = """"""

    def test_part_1(self):
        public_card = 5764801
        public_door = 17807724
        self.assertEqual(solve_part_1(public_card, public_door), 14897079)


if __name__ == '__main__':
    print('Day 25')
    start = time.time()
    public_card = 8335663
    public_door = 8614349

    print('\tPart 1: {}'.format(solve_part_1(public_card, public_door)))

    print("\tTime: {:.4f}ms".format((time.time() - start) * 1000))
    unittest.main()

File: rust/src/day25.rs
const PRIME: u64 = 20201227;

fn solve_part_1(public_card: u64, public_door: u64) -> u64 {
    let mut value = 1;
    let mut encrypt = 1;
    while value != public_card {
        value = value * 7 % PRIME;
        encrypt = encrypt * public_door % PRIME;
    }

    encrypt
}

pub fn main() {
    println!("Day 25");

    let public_card = 8335663;
    let public_door = 8614349;
    println!("\tPart1: {}", solve_part_1(public_card, public_door));
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_part_1() {
        let public_card = 5764801;
        let public_door = 17807724;
        assert_eq!(solve_part_1(public_card, public_door), 14897079);
    }
}
